src/compiler/crystal/types.cr

```diff
@@ -1321,9 +1328,22 @@ module Crystal
   # but can also match other types (like ints and floats) if the literal
   # fits in those types.
   class NumberLiteralType < Type
+    # The literal associated with this type
     getter literal : NumberLiteral
+
+    # The most exact match type, or the first match otherwise
     @matched_type : Type?
 
+    # A type that signals it's an ambiguous call, because the literal
+    # can match two types (for example when 1 matches both Int8 and Int16
+    # restrictions, but there's no Int32 restriction)
+    @ambiguous_call : Type?
+
+    # True if this literal matched against the exact type (for example 1
+    # matched an Int32 restriction). If this is true, there's never an
+    # ambiguous call (so `@ambiguous_call` will be `nil`).
+    @exact_match = false
```

Comment:
If `@exact_match == @ambiguous_call.nil?` always, then it should be a function directly. or just remove it since the usages are already guarded by `@ambiguous_call` been not `nil`.

Reply:
Not necessarily. Say you have the literal `1` and these overloads:
1. Int8: it matches, to `@matched_type` becomes `Int8`
2. Int16: it matches, so `@ambiguous_call` becomes `Int16`
3. Int32: it matches exactly... right now we set `@matched_type` to it, `@ambiguous_call` to `nil` and `@exact_match` to `true`...
4. Int64: it matches, but since `@exact_match` is true, we don't care anymore

I'm thinking how to do it without the `@exact_match` variable, if that's what you are suggesting. I think we can do it by comparing `@matched_type == literal.type`, but we'd have to do that check all the time. Maybe it's not that terrible because it's just a reference comparison.

Is this the change you mean?